### DISCUSSION TITLE
types for `Focusable` wrappers?

### DISCUSSION THREAD

**Discussion by mayank99** [ariakit/ariakit#4029] [Questions] [answered]
Hello! I'm trying to create a `<Textarea>` component on top of [`Focusable`](https://ariakit.org/components/focusable). From what I can tell, this is the proper way to go about it:

```tsx
interface TextareaProps extends Ariakit.FocusableProps<'textarea'> {}

export const Textarea = React.forwardRef<
  React.ElementRef<typeof Ariakit.Focusable>,
  TextareaProps
>((props, forwardedRef) => {
  return (
    <Ariakit.Focusable render={<textarea />} {...props} ref={forwardedRef} />
  );
});
```

I'm not sure if I'm doing something wrong or if this is a bug, but I get this TS error:

```
Types of property 'onChange' are incompatible.
 …
  Types of property 'target' are incompatible.
    Type 'EventTarget' is not assignable to type 'EventTarget & HTMLTextAreaElement'.
      Type 'EventTarget' is missing the following properties from type 'HTMLTextAreaElement': autocomplete, cols, defaultValue, dirName, and 322 more.
```

See [minimal reproduction](https://stackblitz.com/edit/vitejs-vite-jycj9v?file=src%2FApp.tsx&terminal=dev).

It's worth noting that if I change all instances of `textarea` in the code above to `input`, then I don't get any TS errors. So it's possible that `Ariakit.FocusableProps<'textarea'>` doesn't work as intended.

Any help is appreciated!

**Comment by benrodrs** [comment]
I understand this may not be very intuitive, but you should pass the props to a `Role.textarea` component first:

```tsx
export const Textarea = React.forwardRef<
  React.ElementRef<'textarea'>, // <----
  TextareaProps
>((props, forwardedRef) => {
  return (
    <Ariakit.Role.textarea // <---
      {...props}
      ref={forwardedRef}
      render={<Ariakit.Focusable render={<textarea />} />} // <---
    />
  );
});
```

**Reply by mayank99** [reply, marked as answer]
**Update**: PR #4032 has been merged, adding the [`focusable-textarea` example](https://github.com/ariakit/ariakit/blob/main/examples/focusable-textarea/index.tsx). 🚀

---

**Original comment**:

Thank you for the response. My initial version did have `<Ariakit.Role.textarea>` with `render={<textarea />}`. Then I changed it to `<Ariakit.Focusable>` because it seemed more appropriate.

One thing to keep in mind when nesting `render` props to combine `Role.textarea` and `Focusable` is that we don't want to lose out on the ability to accept external `render` props. e.g. This still needs to be supported:

```tsx
<Textarea render={<MyTextarea />} />
```

So I guess the correct way would be like this:

```tsx
<Ariakit.Role.textarea
  {...props}
  ref={forwardedRef}
  render={<Ariakit.Focusable render={props.render || <textarea />} />}
/>
```

Would it make sense to document this? Or have examples for basic `<Input>` and `<Textarea>` components? I'd be happy to contribute if this sounds reasonable.

**Reply by benrodrs** [reply]
Yes, I think that could be a nice `Focusable` example.

**Reply by mayank99** [reply]
Actually I found an issue with this technique. `Focusable` props are being passed to the wrong component. For example, `accessibleWhenDisabled` and `autoFocus` should be spread on `Focusable` but are being spread on `Role.textarea` instead. **Edit**: Please disregard, see https://github.com/ariakit/ariakit/discussions/4029#discussioncomment-10327531.

Whereas in the very original version, there is only a TS issue but it works fine at runtime.

**Reply by mayank99** [reply]
**Edit**: Please disregard.

Maybe `Focusable` should accept a generic type similar to `FocusableProps`?

<details><summary>Diff</summary>

In [`focusable.tsx#L457`](https://github.com/ariakit/ariakit/blob/bf6590a7961bbaf035181c963301150c31642b39/packages/ariakit-react-core/src/focusable/focusable.tsx#L457):

```diff
- export const Focusable = forwardRef(function Focusable(props: FocusableProps) { 
+ export const Focusable = forwardRef(function Focusable<T>(props: FocusableProps<T>) { 
```

</details> 

This would allow using it like this:

```tsx
<Ariakit.Focusable<'textarea'> render={<textarea />} {...props} ref={forwardedRef} />
```

---

Alternatively, maybe `useFocusable` could be exported publicly? This would simplify consumer code while still using `Role.textarea`.

```tsx
<Ariakit.Role.textarea {...Ariakit.useFocusable(props)} ref={forwardedRef} />
```

**Reply by diegohaz** [reply]
> Actually I found an issue with this technique. `Focusable` props are being passed to the wrong component. For example, `accessibleWhenDisabled` and `autoFocus` should be spread on `Focusable` but are being spread on `Role.textarea` instead.

What exactly is the issue? `accessibleWhenDisabled` and `autoFocus` should be passed down to `Focusable` without any problems.

**Reply by mayank99** [reply]
Ahh you're right, it's working correctly (tested on [stackblitz](https://stackblitz.com/edit/vitejs-vite-in5cvx?file=src%2FApp.tsx&terminal=dev)). I was trying to click the textarea and it wasn't being focused, which I think is intended (?). Tabbing to it with my keyboard works fine.

Apologies for the confusion!

**Reply by diegohaz** [reply]
Ah, yes! That makes sense. No worries.
